Wikipedia/Code/SavedArticlesFetcher.m

```diff
@@ -48,39 +43,64 @@ - (void)dealloc {
 }
 
 - (instancetype)initWithDataStore:(MWKDataStore *)dataStore
-                    savedPageList:(MWKSavedPageList *)savedPageList
                    articleFetcher:(WMFArticleFetcher *)articleFetcher
                   imageController:(WMFImageController *)imageController
                  imageInfoFetcher:(MWKImageInfoFetcher *)imageInfoFetcher {
     NSParameterAssert(dataStore);
-    NSParameterAssert(savedPageList);
     NSParameterAssert(articleFetcher);
     NSParameterAssert(imageController);
     NSParameterAssert(imageInfoFetcher);
     self = [super init];
     if (self) {
+        self.fetchesInProcessCount = @0;
         self.accessQueue = dispatch_queue_create("org.wikipedia.savedarticlesarticleFetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
         self.fetchOperationsByArticleTitle = [NSMutableDictionary new];
+
+        [self updateFetchesInProcessCount];
+
         self.errorsByArticleTitle = [NSMutableDictionary new];
         self.dataStore = dataStore;
         self.articleFetcher = articleFetcher;
         self.imageController = imageController;
-        self.savedPageList = savedPageList;
         self.imageInfoFetcher = imageInfoFetcher;
         self.spotlightManager = [[WMFSavedPageSpotlightManager alloc] initWithDataStore:self.dataStore];
+        self.savedArticlesFetcherProgressManager = [[SavedArticlesFetcherProgressManager alloc] initWithDelegate:self];
     }
     return self;
 }
 
-- (instancetype)initWithDataStore:(MWKDataStore *)dataStore
-                    savedPageList:(MWKSavedPageList *)savedPageList {
+- (instancetype)initWithDataStore:(MWKDataStore *)dataStore {
     return [self initWithDataStore:dataStore
-                     savedPageList:savedPageList
                     articleFetcher:[[WMFArticleFetcher alloc] initWithDataStore:dataStore]
                    imageController:[WMFImageController sharedInstance]
                   imageInfoFetcher:[[MWKImageInfoFetcher alloc] init]];
 }
 
+#pragma mark - Progress
+
+// Reminder: due to the internal structure of this class and how it is presently being used, we can't simply check the 'count' of 'fetchOperationsByArticleTitle' dictionary for the total. (It doesn't reflect the actual total.) Could re-plumb this class later.
+- (NSUInteger)calculateTotalArticlesToFetchCount {
+    NSAssert([NSThread isMainThread], @"Must be called on the main thread");
```

Comment:
It doesn't really matter as it's just an assert but it may be of interest to you, I know I was surprised:

[I think it should say](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontextconcurrencytype/nsmainqueueconcurrencytype) "Must be called on the main queue" not "thread" to be precise. Queue != Thread

Queues reuse threads, so it may happen that a background queue will execute on a main thread and [apparently it's very easy to reproduce](http://blog.benjamin-encz.de/post/main-queue-vs-main-thread/).

[More on that topic](http://blog.krzyzanowskim.com/2016/06/03/queues-are-not-bound-to-any-specific-thread/)

As I said, not that it matters or produces any problem but just a little brain teaser 😁

Reply:
Oh thanks! Good catch!

Reply:
That `main-queue-vs-main-thread` article is great!!!